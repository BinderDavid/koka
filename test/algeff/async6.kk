// --------------------------------------------------------
// Async demo: do concurrently
// --------------------------------------------------------
effect amb {
  flip() : bool
}
 
val amb = handler {
  return x -> [x]
  flip()   -> resume(False) + resume(True)
}

function dowait(s : int) {
  val secs = s.show + " seconds"
  println("and waiting at least " + secs)
  wait(s * 1000)
  println("(done waiting " + secs + ")")
  "waited " + secs
}

function test() {
  println("starting")
  function f() {
    println("what is your name?")
    readline()
    error("ouch")
  }
  function g() {
    wait(1000)
    if (flip()) then dowait(2) else dowait(3)
  }
  val xs = asyncAll( [f,g] )
  println("hi " + xs.join(","))  
}

function main() {
  async{ amb(test); () }
}