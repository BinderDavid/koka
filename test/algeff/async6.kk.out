.Eff-amb : forall<a> (.ops-amb<a>) -> operation<a>
.Op-flip : forall<a> .ops-amb<a>
.copy : forall<a> (.ops-amb<a>) -> .ops-amb<a>
amb : forall<a,e> (action : () -> <amb|e> a) -> e list<a>
dowait : (s : int) -> <asyn,console> string
flip : () -> amb bool
main : () -> <exn,console> ()
test : () -> <amb,exn,asyn,console> ()